machines: Allow detaching disks only of either running or shut off VMs

KKoukiou opened the pull request:
Live detaching disks needs the cooperation of the operating system of
the guest to succeed. When trying to live detach disks from paused VMs
the DetachDevice API will timeout and an error will appear in the
cockpit-machines notification list right after.

Let's disable the detach button for VMs that are neither stopped nor
running since the detach operation on these is not likely to succeed.

Relevant to https://bugzilla.redhat.com/show_bug.cgi?id=1708515

KKoukiou commented:
![detach-paused](https://user-images.githubusercontent.com/14921356/57527816-eb035780-7330-11e9-9af5-9952a33d05c3.png)

martinpitt commented:
[These](https://209.132.184.41:8493/logs/pull-11790-20190510-121616-e29f8147-verify-debian-testing/log.html#258) [two failures](https://209.132.184.41:8493/logs/pull-11790-20190510-121616-e29f8147-verify-ubuntu-stable/log.html#245) look unfamiliar:

    RuntimeError: #vm-subVmTestCreate17-state not found

Retrying for comparison.

martinpitt commented:
[This failure](https://209.132.184.41:8493/logs/pull-11790-20190510-143838-e29f8147-verify-debian-testing/log.html#260) looks real, sorry. Check the [screenshot](https://209.132.184.41:8493/logs/pull-11790-20190510-143838-e29f8147-verify-debian-testing/TestMachinesDBus-testDetachDisk-debian-testing-127.0.0.2-2601-FAIL.png).

KKoukiou commented:
> [This failure](https://209.132.184.41:8493/logs/pull-11790-20190510-143838-e29f8147-verify-debian-testing/log.html#260) looks real, sorry. Check the [screenshot](https://209.132.184.41:8493/logs/pull-11790-20190510-143838-e29f8147-verify-debian-testing/TestMachinesDBus-testDetachDisk-debian-testing-127.0.0.2-2601-FAIL.png).

You 're right, it's because suspending VMs require the cooperation of the guest OS, and maybe the image was not booted when I tried to suspend it. Added some wait condition, for it to fully boot.

Antique commented on pkg/machines/components/diskRemove.jsx:
```diff
@@ -37,8 +38,28 @@ const onDetachDisk = (dispatch, vm, target, onAddErrorNotification) => {
+        <OverlayTrigger
+            overlay={
+                <Tooltip id="tip-inforec">
+                    {cockpit.format(_("Detaching disks is not supported for $0 VMs"), vm.state)}
```
The "not supported" is a bit misleading, it is supported but not available/possible with the current VM state.  I'm wondering whether we want to have such check in cockpit, there is nothing like that in virt-manager.  The reason is that even if the VM is running it can still have broken OS or no OS at all and in that case you will not be able to live-detach the disk as well.

KKoukiou replied:
@Antique, thanks for the feedback, I was thinking as well if this check has it's place in cockpit. Since libvirt knows that detaching disks with the live flag from paused VMs will fail anyway, maybe it should return error instead of timeing out.

So, either here or in libvirt the check should exist I believe, since  that pausing will timeout for sure. And for running Vms, where as you said it might fail if OS is broken, we 'll get the fail notification anyway, after the timeout expires. It's just a best effort to prevent the user from trying stupid things.

Do you have some better suggestion? 

PS: I would like to avoid to introduce the logic of persistent XMLs vs live XMLs here as we have for example in the NIC configuration.

That all said I think the check should be somewhere for not allowing the operation, if the place is not cockpit but libvirt, I am ok with that as well.

Antique replied:
It's not that simple, yes we have a timeout in libvirt after witch libvirt returns from the API with success but the device is not actually removed.  The complex part is that the VM OS can just postpone the operation and may remove the device once the timeout expired so it will eventually be removed.  Even with the paused VM it's not that simple, once the VM is resumed the disk will be eventually removed as well.

Probably the only sensible solution in UI is to model it in a way that it will not guarantee the removal but simply "asks" the disk to be removed and it may happen or not.  If the removal is successful libvirt will generate event VIR_DOMAIN_EVENT_ID_DEVICE_REMOVED.  Even if the API returns immediately there is no guarantee that the disk will be actually removed as the VM OS can simply ignore that request.

So the check probably cannot live anywhere and the UI has to inform the user about the conditions and that it is only best-effort operation.

KKoukiou replied:
@Antique I don't know how we can expose such information to the user nicely, expressing that the X disk might get detached, but it might not.

I would like to keep it as deterministic as possible, even with the cost of removing some functionality, aka detaching disks from paused VMs, and hoping that the operation will take efffect after resume.
 Of course for the current patch, the tool tip wording needs fixing.

That said, I agree that I should enhance the detach feature, so that for running VMs it will inform the user that the operation is a best effort thing.

But I still think that this patch should stay, I tend to want to disallow detaching disks from paused VMs since it can give us only headaches.
@martinpitt do you have any feedback on this?

martinpitt replied:
Not a strong one, perhaps @garrett can chime in here. My gut feeling is that it's nice and cheap for the cases where it's definitively not working (paused). Then the tooltip should say "not possible" instead of "not supported". For the running case, the button should maybe say "Request detaching"?

KKoukiou replied:
Or show some notifications about the progress of the operation. (cockpit-composer does this actually)

First we can show a info notification that disk detach has started which will expire and disappear after X seconds. In the mean time we should show an in progress' icon, while waiting for the DEVICE_REMOVED signal. If we don't receive the signal the API fail with timeout, we just show the timeout fail notification and remove the 'in progress' icon.

But in any case, let's not introduce this logic for paused VMs. I am happy to implement it (either this or change the button to 'Request detaching' for the running.

Antique replied:
The API will never fail with timeout error, in this case it will always return success.

As a note, is it OK to remove the 'in progress' icon once the API returns and no DEVICE_REMOVED signal arrived? The signal may arrive in a few seconds after that happen but it can also never arrive or in several minutes/hours.

KKoukiou commented:
This should is blocked till https://github.com/cockpit-project/cockpit/pull/10547 gets merged, since we need the disk change from ide -> virtio to be able to wait for the VM to boot for RHEL hosts.

garrett commented:
Hey! So I've just seen this PR now, due to being tagged right before the weekend.

We could handle this using a few different approaches (ordered from simple to complex changes):

1. Reword to "Disks cannot be removed from paused VMs"

2. Flip the error message into a solution... or include a solution with the message:
   - "Unpause to remove this disk"
   - "Disks cannot be removed from paused VMs. Unpause before detatching this disk."

3. Remove or hide the [-] button while the machine is paused.
   - This can be as simple as using CSS like `.vm-paused .vm-disk-remove { display: none; }`, or just not including the action when building the row in React
   - There's a drawback: This approach may make someone ask "Why can't I remove the disk?" and not realize it's because the VM is paused. 

4. Unpause, remove disk (...and optionally re-pause, although this is probably weird)?
   - This approach lets someone remove a disk regardless of state, and alters state to be able to do so. It's doing whatever it takes to "Make it so" without having to care about the details of why or how.
   - As it is probably not obvious, perhaps a bit surprising, and a little heavy-weight, we could have something like a confirmation dialog to address these concerns. Something to the effect of:

     > Removing $DISK requires an active or stopped VM.
     >
     > [Cancel] [Unpause & Remove Disk]

5. Queue up the action for when the VM resumes. (This was mentioned above.)
   - Similar in spirit to the above, it's another "Make it so" approach. But then we have to handle the state, and it introduces additional edge cases, such as:
      - What happens when the Cockpit session ends before the task runs?
      - What happens on host reboot?

Options 1 & 2 rely on a tooltip to communicate why something isn't possible and make the user perform an additional task. Option 3 sidesteps everything. Options 4 & 5 let the user perform what they want to do and delegate the technicalities to the computer.

Generally, I think having the computer perform necessary tasks is the right approach, but it often means more complexity in the implementation to smooth over the rough edges. But it's better for developers and the software to take care of things instead of forcing the user to do so. That said, this PR does already improve things, so I think it's acceptable to reword the tooltip, merge this PR, and consider possibly implementing something like option 4 or 5 in some future follow-up PR.

KKoukiou commented:
@martinpitt I just fixed the wording of the tooltip according to @garrett 's suggestion and rebased (no conflicts whatsoever). Can you please reapprove?

KKoukiou commented:
Rebased and re-pushed for recently merged flake fixes to be applied.

KKoukiou commented:
Clean from cockpit-machines failures.
